File: Leetcode/1466 Reorder Routes to Make All Paths Lead to the City Zero.cpp
FORWARD, BACKWARD = 1, 0

class Solution:
    def minReorder(self, n, connections):
        self.adj_list = [[] for _ in range(n)]
        for src, dst in connections:
            self.adj_list[src].append((dst, FORWARD))
            self.adj_list[dst].append((src, BACKWARD))
        
        self.n_parent_to_child_edges = 0    
        self.dfs(0, None)
        return self.n_parent_to_child_edges
    
    def dfs(self, root, parent):
        for child, direction in self.adj_list[root]:
            if child != parent:
                self.n_parent_to_child_edges += direction
                self.dfs(child, root)

File: Leetcode/1466 Reorder Routes to Make All Paths Lead to the City Zero.ts
const [FORWARD, BACKWARD] = [1, 0];

function minReorder(n: number, connections: number[][]): number {
    const adjList: [number, number][][] = new Array(n);
    for (let i = 0; i < n; i++) {
        adjList[i] = [];
    }
    for (const [src, dst] of connections) {
        adjList[src].push([dst, FORWARD]);
        adjList[dst].push([src, BACKWARD]);
    }
    let nParentToChildEdges = 0;
    const dfs = (root: number, parent: number) => {
        for (const [child, direction] of adjList[root]) {
            if (child !== parent) {    
                nParentToChildEdges += direction;
                dfs(child, root);
            }
        }
    }
    dfs(0, -1);
    return nParentToChildEdges;
}
